### PR TITLE
New-CmAzDeployment -  initial implementation

### LINKED ISSUE #91: General - Self-loading configuration files
Currently we have to remember which configuration file is associated with each CmdLet. However, we can simplify our deploy scripts considerably with a single deployment command and parameter files which indicate the appropriate provisioning mechanism. Example file named `mycomponent.yml`:

``` yaml
component: core.keys
# rest of config goes here
```

This could then be deployed using a much simpler command, along the lines of:

``` powershell
New-CmAzDeployment -SettingsFile mycomponent.yml
```

Within the command we include what is effectively a large switch statement:

``` powershell
Switch (settings.component)
{
  "core.keys" { New-CmAzCoreKeyVault -SettingsObject settings }
}
```

As well as simpler usage in deployment scripts, this change would lay the groundwork for being able to auto-deploy based on static settings files alone, massively reducing the need for scripting.

- [x] Add "component" to each settings file created via New-CmAzProject, with the appropriate value: Keys, VM etc.
- [x] Create new cmdlet called New-CmAzDeplotment, this should take the settings file as the parameter.
- [x] Create a switch statement that checks the value of the component property in the settings file, and calls the appropriate 
cmdlet. If the component cannot be found, throw an error highlighting this.
- [x] Write a pester test that determines that the correct cmdlet is called for each component. Test the output in the pipeline as well (Use WhatIf and WhatIfPreference).
- [x] Write a pester test checking that an error is thrown when the component is not found, test that the correct error message is displayed as well.

### PR DESCRIPTION
**Pull request by deepend-dev**
closes #91

### CODE REVIEW

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		 Set tags on resources.
+
+		.Description
+		 Sets mandatory and custom tags on either specific resources or all the resources in a resource group.
+
+		.Parameter SettingsFile
+		 File path for the settings file to be converted into a settings object.
+
+		.Parameter SettingsObject
+         Object containing the configuration values required to run this cmdlet.
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:14]: Are these aligned in code?

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	#>
+
+	[CmdletBinding(SupportsShouldProcess, ConfirmImpact = "Medium")]
+	param(
+		[parameter(Mandatory = $true, ParameterSetName = "Settings File")]
+		[String]$SettingsFile,
+		[parameter(Mandatory = $true, ParameterSetName = "Settings Object")]
+        [Object]$SettingsObject,
+        [String]$TagSettingsFile,
+        [parameter(Mandatory = $false)]
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:46]: Are these aligned in code?

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	try {
+
+			if ($SettingsFile -and !$SettingsObject) {
+				$SettingsObject = Get-CmAzSettingsFile -Path $SettingsFile
+			}
+			elseif (!$SettingsFile -and !$SettingsObject) {
+				Write-Error "No valid input settings." -Category InvalidArgument -CategoryTargetName "SettingsObject"
+			}
+
+            switch ($SettingsObject.component) {
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:63]: is this case insensitive?

File: Cloudmarque.Azure/Private/Utility/Get-Prerequisites.ps1 [hunk 4]
```diff
@@ -0,0 +1,10 @@
+function Get-Prerequisites() {
+
+    param(
+        [parameter(Mandatory = $true)]
+        [string]$CommandName
+    )
+
+    Write-Verbose "Invoking $CommandName"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Private/Utility/Get-Prerequisites.ps1:8]: Can we add more info from Get-AzContext to "Invoking $CommandName"?

File: Cloudmarque.Azure/Private/Utility/Get-Prerequisites.ps1 [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+    Get-CmAzSubscriptionName
```
Comment (tm-luke) [inline Cloudmarque.Azure/Private/Utility/Get-Prerequisites.ps1:9]: I think this should be separate from the command, as Get-Prerequisites should output invoke information

**Review by tm-luke — CHANGES_REQUESTED** [review]
Initial review, will look more after.

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 (hunk 2 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:46]: yes

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 (hunk 3 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:63]: yes

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Private/Utility/Get-Prerequisites.ps1 (hunk 4 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Private/Utility/Get-Prerequisites.ps1:8]: we decided on friday that we will just do verbose message - "invoke command" only

**Pull request closed without merging**